### PR TITLE
explain: add subqueries

### PR DESCRIPTION
**Pull request by Joyinqin**
### What is changed, added or deleted? (Required)
This is a followup PR to https://github.com/pingcap/docs/pull/3858

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [x] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/docs/pull/3948
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Have version specific changes 
- [ ] Might cause conflicts

### CODE REVIEW

File: explain-subqueries.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
+---
+title: 用 Explain 查看使用子查询的执行计划
```
Comment (TomShawn) [inline explain-subqueries.md:2]:
```suggestion
title: 用 EXPLAIN 查看使用子查询的执行计划
```

File: explain-subqueries.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+summary: 了解 TiDB 中 EXPLAIN 语句返回的执行计划信息。
+---
+
+# 用 Explain 查看使用子查询的执行计划
```
Comment (TomShawn) [inline explain-subqueries.md:6]:
```suggestion
# 用 EXPLAIN 查看使用子查询的执行计划
```

File: explain-subqueries.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+TiDB 会执行多种[子查询相关的优化](/subquery-optimization.md)，以提升子查询的执行性能。本文档介绍了一些常见的子查询优化方式，以及如何解读 `EXPLAIN` 语句返回的执行计划信息。
```
Comment (TomShawn) [inline explain-subqueries.md:8]:
```suggestion
TiDB 会执行多种[子查询相关的优化](/subquery-optimization.md)，以提升子查询的执行性能。本文档介绍一些常见子查询的优化方式，以及如何解读 `EXPLAIN` 语句返回的执行计划信息。
```

File: explain-subqueries.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+UPDATE t1 SET int_col = 1 WHERE pad1 = (SELECT pad1 FROM t1 ORDER BY RAND() LIMIT 1);
+INSERT INTO t3 SELECT NULL, id FROM t1 WHERE id < 1000;
+
+SELECT SLEEP(1);
+ANALYZE TABLE t1, t2, t3;
+```
+
+## Inner Join（无 `UNIQUE` 约束的子查询）
+
+以下示例中，`IN` 子查询会从表 `t2` 中搜索 id 列。考虑到语义正确性，TiDB 需要保证 `t1_id` 列是唯一的。可以使用 `EXPLAIN` 语句查看用于删除重复项并执行 `Inner Join` 操作的执行计划：
```
Comment (TomShawn) [inline explain-subqueries.md:50]:
```suggestion
以下示例中，`IN` 子查询会从表 `t2` 中搜索一列 ID。为保证语义正确性，TiDB 需要保证 `t1_id` 列具有唯一性。使用 `EXPLAIN` 可查看到该查询的执行计划删除重复项并执行 `Inner Join` 操作：
```

File: explain-subqueries.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+| │   └─HashAgg_31               | 45.00    | cop[tikv] |                              | group by:test.t2.t1_id,                                                         |
+| │     └─IndexFullScan_37       | 90000.00 | cop[tikv] | table:t2, index:t1_id(t1_id) | keep order:false                                                                |
+| └─TableReader_14(Probe)        | 1.00     | root      |                              | data:TableRangeScan_13                                                          |
+|   └─TableRangeScan_13          | 1.00     | cop[tikv] | table:t1                     | range: decided by [test.t2.t1_id], keep order:true                              |
++--------------------------------+----------+-----------+------------------------------+---------------------------------------------------------------------------------+
+7 rows in set (0.00 sec)
+```
+
+由上述查询结果可知，TiDB 首先执行 `Index Join`（`Merge Join` 的变体）操作，开始读取 `t2.t1_id` 列的索引。`t1_id` 的值是 `└─HashAgg_31` 算子任务的一部分，先是在 TiKV 内对其进行重复数据删除；同时该值也是 `├─HashAgg_38(Build)` 算子任务的一部分，在 TiDB 中会再次对其进行重复数据删除。这项重复数据删除操作由聚合函数 `firstrow(test.t2.t1_id)` 执行，之后会将结果与 `t1` 表的 `PRIMARY KEY` 连接在一起。
```
Comment (TomShawn) [inline explain-subqueries.md:71]:
```suggestion
由上述查询结果可知，TiDB 首先执行 `Index Join`（`Merge Join` 的变体）操作，开始读取 `t2.t1_id` 列的索引。先是 `└─HashAgg_31` 算子的部分任务在 TiKV 中对 `t1_id` 值进行去重，然后`├─HashAgg_38(Build)` 算子的部分任务在 TiDB 中对 `t1_id` 值再次进行去重。去重操作由聚合函数 `firstrow(test.t2.t1_id)` 执行，之后会将操作结果与 `t1` 表的主键相连接。
```

**Review by TomShawn — COMMENTED** [review]
round 1

File: explain-subqueries.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+|   └─Selection_21            | 56808.00  | cop[tikv] |                              | ne(test.t1.id, test.t1.int_col)                                                                        |
+|     └─TableFullScan_20      | 71010.00  | cop[tikv] | table:t1                     | keep order:true                                                                                        |
++-----------------------------+-----------+-----------+------------------------------+--------------------------------------------------------------------------------------------------------+
+6 rows in set (0.00 sec)
+```
+
+由上述查询结果可知，TiDB 执行了 `Semi Join`。不同于 `Inner Join`，`Semi Join` 仅认可右键（`t2.t1_id`）上的第一个值，也就是该操作将删除 `Join` 算子任务中的重复数据。`Join` 算法也包含 `Merge Join`，会按照排序顺序同时从左侧和右侧读取数据，这是一种高效的 `Zipper Merge`。
+
+可以认为原语句为*关联子查询*，因为它引入了子查询外的 `t1.int_col` 列。然而，`EXPLAIN` 语句的返回结果显示了应用[关联子查询去关联](/correlated-subquery-optimization.md)后的执行计划。条件 `t1_id != t1.int_col` 会重写为 `t1.id != t1.int_col`。TiDB 可以从表 `t1` 中读取数据并且在 `└─Selection_21` 中执行此操作，因此这种去关联和重写操作会极大提高执行效率。
```
Comment (TomShawn) [inline explain-subqueries.md:122]:
```suggestion
可以将原语句视为*关联子查询*，因为它引入了子查询外的 `t1.int_col` 列。然而，`EXPLAIN` 语句的返回结果显示的是[关联子查询去关联](/correlated-subquery-optimization.md)后的执行计划。条件 `t1_id != t1.int_col` 会被重写为 `t1.id != t1.int_col`。TiDB 可以从表 `t1` 中读取数据并且在 `└─Selection_21` 中执行此操作，因此这种去关联和重写操作会极大提高执行效率。
```

File: explain-subqueries.md [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+## Anti Semi Join （`NOT IN` 子查询）
+
+在以下示例中，*除非*子查询中存在 `t3.t1_id`，否则查询时将从语义上返回表 `t3` 中的所有行：
```
Comment (TomShawn) [inline explain-subqueries.md:126]:
```suggestion
在以下示例中，*除非*子查询中存在 `t3.t1_id`，否则该查询将（从语义上）返回表 `t3` 中的所有行：
```

File: explain-subqueries.md [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+| ├─TableReader_28(Build)     | 1998.00 | root      |               | data:TableFullScan_27                                                               |
+| │ └─TableFullScan_27        | 1998.00 | cop[tikv] | table:t3      | keep order:false                                                                    |
+| └─TableReader_15(Probe)     | 1.00    | root      |               | data:Selection_14                                                                   |
+|   └─Selection_14            | 1.00    | cop[tikv] |               | lt(test.t1.int_col, 100)                                                            |
+|     └─TableRangeScan_13     | 1.00    | cop[tikv] | table:t1      | range: decided by [test.t3.t1_id], keep order:true                                  |
++-----------------------------+---------+-----------+---------------+-------------------------------------------------------------------------------------+
+6 rows in set (0.00 sec)
+```
+
+上述查询首先读取了表 `t3`，然后根据主键开始查询表 `t1`。`Join` 类型是 _anti semi join_：之所以使用 _anti_，是因为上述示例有不存在匹配值（即 `NOT IN`）的情况；`Semi Join` 则是因为仅需要匹配第一行后就可以停止查询。
```
Comment (TomShawn) [inline explain-subqueries.md:146]:
```suggestion
上述查询首先读取了表 `t3`，然后根据主键开始探测 (probe) 表 `t1`。连接类型是 _anti semi join_：之所以使用 _anti_，是因为上述示例有不存在匹配值（即 `NOT IN`）的情况；使用 `Semi Join` 则是因为仅需要匹配第一行后就可以停止查询。
```

File: explain-subqueries.md [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+| │ └─IndexFullScan_38        | 999.00  | cop[tikv] | table:t3, index:t1_id(t1_id) | keep order:false                                                                |
+| └─TableReader_15(Probe)     | 1.00    | root      |                              | data:TableRangeScan_14                                                          |
+|   └─TableRangeScan_14       | 1.00    | cop[tikv] | table:t1                     | range: decided by [test.t3.t1_id], keep order:true                              |
++-----------------------------+---------+-----------+------------------------------+---------------------------------------------------------------------------------+
+5 rows in set (0.00 sec)
+```
+
+从语义上看，因为保证了 `t3.t1_id` 的唯一性，TiDB 可以直接执行 `INNER JOIN` 查询。
+
+## Semi Join（关联查询）
```
Comment (TomShawn) [inline explain-subqueries.md:96]:
```suggestion
## Semi Join（半连接关联查询）
```

File: explain-subqueries.md [hunk 10]
```diff
@@ -0,0 +1,146 @@
+---
+title: 用 EXPLAIN 查看使用子查询的执行计划
+summary: 了解 TiDB 中 EXPLAIN 语句返回的执行计划信息。
+---
+
+# 用 EXPLAIN 查看使用子查询的执行计划
+
+TiDB 会执行多种[子查询相关的优化](/subquery-optimization.md)，以提升子查询的执行性能。本文档介绍一些常见子查询的优化方式，以及如何解读 `EXPLAIN` 语句返回的执行计划信息。
+
+本文档所使用的示例数据如下:
```
Comment (TomShawn) [inline explain-subqueries.md:10]:
```suggestion
本文档所使用的示例表数据如下:
```

File: explain-subqueries.md [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+UPDATE t1 SET int_col = 1 WHERE pad1 = (SELECT pad1 FROM t1 ORDER BY RAND() LIMIT 1);
+INSERT INTO t3 SELECT NULL, id FROM t1 WHERE id < 1000;
+
+SELECT SLEEP(1);
+ANALYZE TABLE t1, t2, t3;
+```
+
+## Inner Join（无 `UNIQUE` 约束的子查询）
+
+以下示例中，`IN` 子查询会从表 `t2` 中搜索一列 ID。为保证语义正确性，TiDB 需要保证 `t1_id` 列的值具有唯一性。使用 `EXPLAIN` 可查看到该查询的执行计划删除重复项并执行 `Inner Join` 内连接操作：
```
Comment (TomShawn) [inline explain-subqueries.md:50]:
```suggestion
以下示例中，`IN` 子查询会从表 `t2` 中搜索一列 ID。为保证语义正确性，TiDB 需要保证 `t1_id` 列的值具有唯一性。使用 `EXPLAIN` 可查看到该查询的执行计划去掉重复项并执行 `Inner Join` 内连接操作：
```

File: explain-subqueries.md [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+INSERT INTO t2 SELECT NULL, a.id, RANDOM_BYTES(1024), RANDOM_BYTES(1024), RANDOM_BYTES(1024) FROM t1 a JOIN t1 b JOIN t1 c LIMIT 10000;
+INSERT INTO t2 SELECT NULL, a.id, RANDOM_BYTES(1024), RANDOM_BYTES(1024), RANDOM_BYTES(1024) FROM t1 a JOIN t1 b JOIN t1 c LIMIT 10000;
+UPDATE t1 SET int_col = 1 WHERE pad1 = (SELECT pad1 FROM t1 ORDER BY RAND() LIMIT 1);
+INSERT INTO t3 SELECT NULL, id FROM t1 WHERE id < 1000;
+
+SELECT SLEEP(1);
+ANALYZE TABLE t1, t2, t3;
+```
+
+## Inner Join（无 `UNIQUE` 约束的子查询）
```
Comment (TomShawn) [inline explain-subqueries.md:48]:
```suggestion
## Inner join（无 `UNIQUE` 约束的子查询）
```

File: explain-subqueries.md [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+|   └─TableRangeScan_14       | 1.00    | cop[tikv] | table:t1                     | range: decided by [test.t3.t1_id], keep order:true                              |
++-----------------------------+---------+-----------+------------------------------+---------------------------------------------------------------------------------+
+5 rows in set (0.00 sec)
+```
+
+从语义上看，因为约束保证了 `t3.t1_id` 列值的唯一性，TiDB 可以直接执行 `INNER JOIN` 查询。
+
+## Semi Join（关联查询）
+
+在前两个示例中，子查询数据的唯一性通过 `HashAgg` 聚合操作或通过 `UNIQUE` 约束保证之后，TiDB 才能够执行 `Inner Join` 操作。这两种连接均使用了 `Index Join`（`Merge Join` 的变体）。
```
Comment (TomShawn) [inline explain-subqueries.md:98]:
```suggestion
在前两个示例中，通过 `HashAgg` 聚合操作或通过 `UNIQUE` 约束保证子查询数据的唯一性之后，TiDB 才能够执行 `Inner Join` 操作。这两种连接均使用了 `Index Join`（`Merge Join` 的变体）。
```

File: explain-subqueries.md [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+| │ └─IndexFullScan_23        | 180000.00 | cop[tikv] | table:t2, index:t1_id(t1_id) | keep order:true                                                                                        |
+| └─TableReader_22(Probe)     | 56808.00  | root      |                              | data:Selection_21                                                                                      |
+|   └─Selection_21            | 56808.00  | cop[tikv] |                              | ne(test.t1.id, test.t1.int_col)                                                                        |
+|     └─TableFullScan_20      | 71010.00  | cop[tikv] | table:t1                     | keep order:true                                                                                        |
++-----------------------------+-----------+-----------+------------------------------+--------------------------------------------------------------------------------------------------------+
+6 rows in set (0.00 sec)
+```
+
+由上述查询结果可知，TiDB 执行了 `Semi Join`。不同于 `Inner Join`，`Semi Join` 仅认可右键（`t2.t1_id`）上的第一个值，也就是该操作将删除 `Join` 算子任务中的重复数据。`Join` 算法也包含 `Merge Join`，会按照排序顺序同时从左侧和右侧读取数据，这是一种高效的 `Zipper Merge`。
```
Comment (TomShawn) [inline explain-subqueries.md:120]:
```suggestion
由上述查询结果可知，TiDB 执行了 `Semi Join`。不同于 `Inner Join`，`Semi Join` 仅允许右键 (`t2.t1_id`) 上的第一个值，也就是该操作将去除 `Join` 算子任务中的重复数据。`Join` 算法也包含 `Merge Join`，会按照排序顺序同时从左侧和右侧读取数据，这是一种高效的 `Zipper Merge`。
```

**Comment by TomShawn** [comment]
LGTM

**Review by lzmhhh123 — APPROVED** [review]
LGTM

**Review by TomShawn — APPROVED** [review]
LGTM

**Pull request merged**